### PR TITLE
Remove the "links" field from the search index

### PR DESCRIPTION
**Pull request by seanh**
Remove the "links" field of annotations from the search index.

This field did not exist in the search index back when we used the
legacy Elasticsearch storage, it only existed in the annotations
returned by the API, but it got accidentally added into the search index when
we moved to the new Postgres storage becayse the new search indexer uses
AnnotationJSONPresenter which is the presenter used for returning
annotations from the API.

Fundamentally we don't want the annotations returned by the API to have
to look exactly the same as the annotations in the search index - these
are two different things, and may evolve in different directions
according to different needs.

So add a new AnnotationSearchIndexPresenter and have the search indexer
use that instead of AnnotationJSONPresenter. Now the annotation formats
used by the API and search index are separated in the code and can be
modified independently.

Note that both AnnotationJSONPresenter and
AnnotationSearchIndexPresenter still use DocumentJSONPresenter for the
document sub-object of annotations - it has not been necessary to
separate out a DocumentSearchIndexPresenter yet.

Fixes https://github.com/hypothesis/h/issues/3261

### CODE REVIEW

File: h/api/presenters.py [hunk 1]
```diff
@@ -90,14 +93,41 @@ def asdict(self):
 
     @property
     def permissions(self):
-        read = self.annotation.userid
-        if self.annotation.shared:
-            read = 'group:{}'.format(self.annotation.groupid)
+        return _permissions(self.annotation)
+
+
+class AnnotationSearchIndexPresenter(AnnotationBasePresenter):
```
Comment (chdorner) [inline h/api/presenters.py]: Some basic testing would be nice here, so we're not going to accidentally break search indexing.

File: h/api/search/index.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def index(es, annotation, request):
     # FIXME: this should not use the same presenter as we use to render
     # annotations for clients. It is useful in the mean time until we get rid of
     # the legacy ElasticSearch annotation storage.
```
Comment (chdorner) [inline h/api/search/index.py]: This comment can go now as well.

**Comment by chdorner** [comment]
Looks good except for a few small things here and there. I just checked how the rendering works in the API and we're still using `annotation_from_dict` (which was not on my list, I will create a new card for that), we just have to remember to still properly present it when rendering.

File: h/api/search/index.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def index(es, annotation, request):
...
-    annotation_dict = presenters.AnnotationJSONPresenter(
+    annotation_dict = presenters.AnnotationSearchIndexPresenter(
         request, annotation).asdict()
 
     annotation_dict['target'][0]['scope'] = [
```
Comment (nickstenning) [inline h/api/search/index.py]: This should probably move into the search index presenter, then?

File: h/api/presenters.py [hunk 4]
```diff
@@ -90,14 +93,41 @@ def asdict(self):
...
+            'created': self.created,
+            'updated': self.updated,
+            'user': self.annotation.userid,
+            'uri': self.annotation.target_uri,
+            'text': self.text,
+            'tags': self.tags,
+            'group': self.annotation.groupid,
+            'permissions': self.permissions,
+            'target': self.target,
+            'document': docpresenter.asdict(),
```
Comment (nickstenning) [inline h/api/presenters.py]: On the basis that we don't index this, I think we can probably remove this now.

File: h/api/presenters.py [hunk 5]
```diff
@@ -90,14 +93,41 @@ def asdict(self):
...
+        }
+
+        if self.annotation.references:
+            base['references'] = self.annotation.references
 
-        return {'read': [read],
-                'admin': [self.annotation.userid],
-                'update': [self.annotation.userid],
-                'delete': [self.annotation.userid]}
+        annotation = copy.copy(self.annotation.extra) or {}
```
Comment (nickstenning) [inline h/api/presenters.py]: We should probably also not be indexing extras, because doing so can result in nasty mapping clashes.

File: h/api/search/index.py (hunk 3 above)
Comment (nickstenning) [inline h/api/search/index.py]: If you do, you can move it in as `uri_normalized` rather than scope and make the appropriate changes to the search query builder.

**Comment by nickstenning** [comment]
I've made a couple of comments, not all of which are suggesting changes to this PR, but changes which are made possible/desirable as a result of doing this work. I think for me all I'd like to see is a few tests for this and then we can merge and follow up on the other stuff I've mentioned.

File: h/api/search/index.py [hunk 6]
```diff
@@ -147,7 +147,7 @@ def _prepare(self, annotation):
         # FIXME: this should not use the same presenter as we use to render
         # annotations for clients. It is useful in the mean time until we get rid of
         # the legacy ElasticSearch annotation storage.
-        data = presenters.AnnotationJSONPresenter(self.request, annotation).asdict()
+        data = presenters.AnnotationSearchIndexPresenter(self.request, annotation).asdict()
         data['target'][0]['scope'] = [annotation.target_uri_normalized]
```
Comment (nickstenning) [inline h/api/search/index.py]: Ditto above. The reason this is separate is because we didn't want to add this to `AnnotationJSONPresenter`. There's no reason not to add it to `AnnotationSearchIndexPresenter`.

File: h/api/presenters.py (hunk 4 above)
Comment (chdorner) [inline h/api/presenters.py]: Not yet I believe, because the search endpoint still uses `annotation_from_dict` and then presents the response from ES. As soon as that endpoint goes to PG as well after searching, then we can remove this.

File: h/api/presenters.py (hunk 5 above)
Comment (chdorner) [inline h/api/presenters.py]: Same as above, we still need to keep this for a bit.

File: h/api/presenters.py (hunk 4 above)
Comment (nickstenning) [inline h/api/presenters.py]: Ugh, right. Sorry.

File: h/api/presenters.py (hunk 5 above)
Comment (nickstenning) [inline h/api/presenters.py]: :+1:

File: h/api/search/index.py (hunk 3 above)
Comment (seanh) [inline h/api/search/index.py]: I'd like to do the rename in a separate pr

File: h/api/presenters.py (hunk 1 above)
Comment (seanh) [inline h/api/presenters.py]: Tests added

File: h/api/search/index.py (hunk 2 above)
Comment (seanh) [inline h/api/search/index.py]: Done

File: h/api/search/index.py (hunk 6 above)
Comment (seanh) [inline h/api/search/index.py]: Done

**Comment by seanh** [comment]
If this can be merged it'll unblock my work on the new `Annotation` test factory

File: tests/h/api/presenters_test.py [hunk 7]
```diff
@@ -446,3 +562,16 @@ def tzname(self, dt):
...
+def clear_registry():
+    # If you call add_annotation_link_generator() as some tests in this module
+    # do it'll add stuff to the request registry that you pass to it.
+    # Then when a later test runs and creates a _new_ DummyRequest instance,
+    # that request's registry will still contain the stuff added by the
+    # previous test!
+    # DummyRequest.registry is shared across requests.
+    # So rather than requiring each test that adds to the registry to also
+    # clear it, this autouse fixture just clears the registry after every test.
+    DummyRequest().registry.clear()
```
Comment (nickstenning) [inline tests/h/api/presenters_test.py]: Hmm. I think the cleanup here should probably be on the relevant tests, or global. But ¯\_(ツ)_/¯... works for now.

**Comment by nickstenning** [comment]
LGTM.
